packages/vats/src/transfer.js

```diff
@@ -0,0 +1,186 @@
+import { M } from '@endo/patterns';
+import { coerceToData, dataToBase64 } from '@agoric/network';
+import { AppI } from './bridge-target.js';
+
+const { Fail } = assert;
+
+const ReactionGuard = M.call(M.any()).optional(M.any()).returns(M.any());
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {import('@agoric/vow').VowTools} vowTools
+ */
+export const prepareTransferInterceptor = (zone, vowTools) => {
+  const { watch } = vowTools;
+  const makeTransferInterceptorKit = zone.exoClassKit(
+    'TransferInterceptorKit',
+    {
+      public: AppI,
+      encodeAckWatcher: M.interface('EncodeAckWatcher', {
+        onFulfilled: ReactionGuard,
+      }),
+      sendErrorWatcher: M.interface('SendErrorWatcher', {
+        onRejected: ReactionGuard,
+      }),
+      logErrorWatcher: M.interface('LogErrorWatcher', {
+        onRejected: ReactionGuard,
+      }),
+    },
+    /**
+     * @param {ERef<import('./bridge-target.js').System>} system
+     * @param {ERef<import('./bridge-target.js').App>} tap
+     * @param {boolean} [isActiveTap] Whether the tap is active (can modify
+     *   acknowledgements), or passive (can't cause delays in the middleware).
+     */
+    (system, tap, isActiveTap = false) => ({ isActiveTap, system, tap }),
+    {
+      public: {
+        async upcall(obj) {
+          const { isActiveTap, system, tap } = this.state;
+
+          obj.type === 'VTRANSFER_IBC_EVENT' ||
+            Fail`Invalid upcall argument type ${obj.type}; expected VTRANSFER_IBC_EVENT`;
+
+          // First, call our target contract listener.
+          // A VTransfer active interceptor can return a write acknowledgement
+          /** @type {import('@agoric/vow').Vow<any>} */
+          let retP = watch(E(tap).upcall(obj));
+
+          // See if the upcall result needs special handling.
+          if (obj.event === 'writeAcknowledgement') {
+            const ackMethod = {
+              type: 'IBC_METHOD',
+              method: 'receiveExecuted',
+              packet: obj.packet,
+            };
+            if (isActiveTap) {
+              // FIXME: use the Vow `watch` operator.
+              retP = watch(retP, this.facets.encodeAckWatcher, { ackMethod });
+            } else {
+              // This is a passive tap, so forward the ack without intervention.
+              ackMethod.ack = obj.acknowledgement;
+              retP = watch(E(system).downcall(ackMethod));
+            }
+            retP = watch(retP, this.facets.sendErrorWatcher, { ackMethod });
+          }
+
+          // Log errors in the upcall handling.
+          retP = watch(retP, this.facets.logErrorWatcher, { obj });
+
+          if (isActiveTap) {
+            // If the tap is active, return the promiseVow to the caller.  This
+            // will delay the middleware until fulfilment of the retP chain.
+            return retP;
+          }
+
+          // Otherwise, passively return nothing.
+        },
+      },
+      encodeAckWatcher: {
```

Comment:
Where are these methods actually explained?

Reply:
Added some comments above the `*Watcher` facets.